services/venice-server/src/main/java/com/linkedin/venice/listener/StorageReadRequestsHandler.java

```diff
@@ -428,10 +433,8 @@ private CompletableFuture<ReadResponse> handleMultiGetRequestInParallel(
     CompletableFuture[] chunkFutures = new CompletableFuture[splitSize];
     PartitionerConfig partitionerConfig = getPartitionerConfig(request.getResourceName());
 
-    Optional<List<Integer>> optionalKeyList =
-        keyValueProfilingEnabled ? Optional.of(new ArrayList<>(totalKeyNum)) : Optional.empty();
-    Optional<List<Integer>> optionalValueList =
-        keyValueProfilingEnabled ? Optional.of(new ArrayList<>(totalKeyNum)) : Optional.empty();
+    IntList responseKeySizeList = keyValueProfilingEnabled ? new IntArrayList() : null;
+    IntList responseValueSizeList = keyValueProfilingEnabled ? new IntArrayList() : null;
```

Comment:
Let's pass `totalKeyNum` in the `IntArrayList` constructor, since we know the size to allocate ahead of time.

Besides efficiency, more importantly, this is required for correctness since we want to set by index from many threads concurrently, and not have the list resize under us.

Reply:
Good catch. add(index, value) would work without proper initialization. But using `set` would throw exceptions.